src/main/kotlin/org/jitsi/nlj/MediaSourceDesc.kt

```diff
@@ -0,0 +1,162 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.nlj
+
+import org.jitsi.nlj.RtpLayerDesc.Companion.getEncodingId
+import org.jitsi.nlj.rtp.VideoRtpPacket
+import org.jitsi.utils.ArrayUtils
+import java.util.Collections
+
+/**
+ * Represents a collection of [RtpLayerDesc]s that encode the same
+ * media source. This specific implementation provides webrtc simulcast stream
+ * suspension detection.
+ *
+ * We take the definition of "Media Source" from RFC 7656.  It takes a single
+ * logical source of media, which might be represented by multiple encoded streams.
+ *
+ * @author George Politis
+ */
+class MediaSourceDesc
+@JvmOverloads constructor(
+    /**
+     * The [RtpEncodingDesc]s that this [MediaSourceDesc]
+     * possesses, ordered by their subjective quality from low to high.
+     */
+    val rtpEncodings: Array<RtpEncodingDesc>,
+    /**
+     * A string which identifies the owner of this source (e.g. the endpoint
+     * which is the sender of the source).
+     */
+    val owner: String? = null
+) {
+    /**
+     * Current single-list view of all the encodings' layers.
+     */
+    private lateinit var layers: List<RtpLayerDesc>
+
+    /**
+     * Allow the lookup of a layer by the encoding id of a received packet.
+     */
+    private val layersById: MutableMap<Long, RtpLayerDesc> = HashMap()
+
+    /**
+     * Allow the lookup of a layer by index.
+     */
+    private val layersByIndex: MutableMap<Int, RtpLayerDesc> = HashMap()
+
+    /**
+     * Update the layer cache.  Should be synchronized on [this].
+     */
+    private fun updateLayerCache() {
+        layersById.clear()
+        layersByIndex.clear()
+        val layers_ = ArrayList<RtpLayerDesc>()
+
+        for (encoding in rtpEncodings) {
+            for (layer in encoding.layers) {
+                layersById[encoding.encodingId(layer)] = layer
+                layersByIndex[layer.index] = layer
+                layers_.add(layer)
+            }
+        }
+        layers = Collections.unmodifiableList(layers_)
+    }
+
+    init { updateLayerCache() }
+
+    /**
+     * Gets the last "stable" bitrate (in bps) of the encoding of the specified
+     * index. The "stable" bitrate is measured on every new frame and with a
+     * 5000ms window.
+     *
+     * @return the last "stable" bitrate (bps) of the encoding at the specified
+     * index.
+     */
+    fun getBitrateBps(nowMs: Long, idx: Int): Long {
+        val layer = getRtpLayerByQualityIdx(idx) ?: return 0
+
+        // TODO: previous code returned a lower layer if this layer's bitrate was 0.
```

Comment:
This is used in probing and the thinking was, if encoding N isn't streaming, get the bitrate of encoding N-1.

Reply:
Okay.  I feel like this makes sense for the probing but not for the source bitrate allocation?  Should the two usages be separated?  Or do you think it makes sense for the source bitrate allocation as well?

Reply:
I want to answer affirmatively but I'm not confident that it makes sense because bandwidth probing and bandwidth allocation are tightly coupled. That said, we may want to change the underlying implementation; we may, for example, want to return a rolling window max for each encoding or a fixed value taken from the Chrome codebase, or something else, instead of what we currently do which is to take the actual bitrate.